### ISSUE TITLE
Epic: Improve LastWrittenLsn mechanism

### ISSUE THREAD

**Issue by MMeent** [neondatabase/neon#2760]
## Motivation

Currently, we have a limit of N=128 block ranges (or 1028 blocks after #2493) with LRU eviction. This is OK, but we could evict many pages at once, resulting in a relatively high overall evicted LSN.

Because a high LastWrittenLSN means a short WAL propagation time to PageServer, it is better to have the lowest possible (within reason) LastWrittenLSN (or: the lowest reasonably possible LSN in the request). 
Even for pages that haven't been updated in a long time and where the WAL propagation delay is not an issue, a lower LastWrittenLsn will result in higher overall performance because we are able to exclude more layers from the search for WAL in the Pageserver.

### Pairingheap (?) cache eviction

We could change our cache eviction policy to one based on the lowest LSN in the cache, e.g., through a pairing heap on the lowest LSN.
This way, we could have 

```
overallLastWrittenLsn - lowest granularity LSN for all blocks not in the rest of the infra
recentlyEvicted - N entries associated with BufferTag
currentLsn - max LSN we can expect in any block

overallLastWrittenLsn < recentlyEvicted (N entries) < currentLsn
```

This results in more precise LSN lookups for pages where it matters - those with a low LSN.

Note that we use this LSN stuff only to ensure that Pageserver is up-to-date enough, so it makes sense to keep granular data for only the most recently updated evicted pages.

### Probabilistic lastWrittenLsn

We could update our LSN cache to one based on count-min sketches: we maintain N buckets, and for all evicted pages, we hash the page ID to correlate to M (M < N) buckets and update those buckets' lastWrittenLsn value to Max(currentVal, pageLsn). 
When looking up the lastWrittenLsn for a page, we do Min(associated buckets' lastWrittenLsn), which is the actual last written LSN for that page. The benefit is a better estimated last written LSN in less memory than a hash map on all pages (no hashmap-related maintenance infrastructure necessary), and it covers the whole database.
I'll refer to "count-min sketches" and "Counting bloom filters" for more details on why this would work, though here we use "Max(insert_value, current_value)" instead of "current_value++" in the buckets.

This could replace only XLogCtl->maxLastWrittenLsn, or both XLogCtl->maxLastWrittenLsn and the lastWrittenLsnCache. I don't care much about either, but precise data (see Pairingheap cache eviction above) could be useful as a first item, after which this probabilistic structure could provide the next, more granular layer of LSNs.

Note that this probabilistic lastWrittenLsn map could also be used in the pageserver, as it could be used to help determine the age of a block in the cluster.

## Other notes

We should probably do something to pull the LSN tracking code to the extension.

**Comment by knizhnik** [comment]
I think that speaking about last written LSN cache we should not forgot about two things:
1. Shared buffers. Last recently access (and modified) pages are used to be be present in shared buffers for some time. Yes, in case of using ring buffer this time can be quite small. But still probability that recently updated page will be evicted is quite small.
2. Backpressure. Right now lag between pageserver and compute is limited by 15Mb. It is just 2000 pages. And default size of last written LSN cache is 128k entries.

so I do not think that current last written LSN cache can cause some slowdown of Neon - just because `get_page_at_lsn` has to wait until irrelevant changes are applied.

The only exception is prefetch. Here, as we see, last written LSN cache miss + updates of traverse pages (as in vacuum) can cause rejection of prefetch requests. But in this case setting last written LSN for prefetched pages can help. Yes, it will "pollute" cache, but still traversing of 128k pages will take significant amount of time, given a pagesever good chance to proceed 2k  pages

**Comment by knizhnik** [comment]
> I disagree with this calculation. 

Yes, you are certainly right that one WAL page may contain multiple records and so touch multiple page. In case of random updates (pgbench) situation will be close to one you have described (no 44 bytes, but but still ~hundred dirty pages. But  I have considered mostly COPY+vacuum workload. In this case mostly FPI are used and number of  modified pages ~= size of WAL.

> I disagree here as well  - cache is currently LRU, so the last evicted value is added and the least recently evicted value is removed. 

Yes, eviction policy is LRU. But first of all "least recently evicted value is removed" is not true. Page is moved to the head of LRU list on each access (update of last written LSN). Last written LSN is set when page is written by SMGR. It doesn't mean that it is evicted. I thought about maintaining really "last evicted LSN" rather than "last wiritten". But it requires non-trivial patch in bufmgr and doesn't give any advantage on any workload  I have tried.

It seems to be useless to dispute here whether 128k cache is enough to avoid pageserver waits or not. I didn't see use cases where pageserver wait time (which can be observed using Prometheus metrics)  becomes limiting factor for performance. This  is what allows me to write that I do not think that we have problem here. If you disagree, please point me on the example which demonstrates it.


> we also need the LwLsn cache to ensure no page was modified and evicted after prefetching i

Yes, and here we really faced with the problem during VACUUM (and with default `shared_buffers=1MB`). But I hope that PR with setting last written LSN for prefetched request solves the problem.  My first thought was the we really need more "accurate" cache here. But then it seems to me that much simpler solution will be enough.

Please notice that the the problem  takes place only with vacuum/bulk update. And scenario with short WAL records and updaing multiple pages by one WAL page is result of random OLTP workload. But in the last case prefetch is not performed at all. And bottleneck most likely will be compute-pageserver roundrip (if working set doesn't fit in compute cache) which server as some kind of packpressure and prevents large gap of current LSN between compute and pageserver.

> For that, we need a good and accurate cache of which the behavior we can reason about. Re-inserting the prefetched pages does work but adds additional (locking) overhead that we shouldn't need. In addition, the write-in-cache system would fail on large enough prefetch distances with enough concurrent connections by thrashing the cache.

You disagree with my arguments, I disagree (partly) with your arguments... 
I do not thick that before trying to implement some more accurate last written LSN cache we should first try create some benchrmark which demonstrates the problem. It is not good to try to optimize something based just on speculations. Prefetch is any case applicable mostly for OLAP queries
and OLAP workoad rarely requires larger number of parallel queries. So one again, your arguments about larger number of concurrent connections performing queries with larger prefetch distance may never happen in real life. Even if we have parallel vacuum, which concurrently process multiple tables, still there will be just few active backends.

**Comment by Bodobolero** [comment]
also see recent discussion in context of TPC-C benchbase benchmark https://databricks.slack.com/archives/C0922DE2Y13/p1753958578911709?thread_ts=1753112675.564979&cid=C0922DE2Y13
